src/main/java/com/czertainly/core/service/cmp/impl/CmpServiceImpl.java

```diff
@@ -0,0 +1,344 @@
+    private X509Certificate recipient;
+    private List<RequestAttributeDto> issueAttributes;
+    private List<RequestAttributeDto> revokeAttributes;
+    private CertificateService certificateService;
+    @Autowired
+    public void setCertificateService(CertificateService certificateService) { this.certificateService = certificateService; }
+
+    // -- CRYPTO
+    private CertificateKeyService certificateKeyService;
+    @Autowired
+    private void setCertificateKeyService(CertificateKeyService certificateKeyService) {  this.certificateKeyService = certificateKeyService; }
+
+    // -- HANDLER
+    private CrmfMessageHandler crmfMessageHandler;
+    @Autowired public void setCrmfMessageHandler(CrmfMessageHandler crmfMessageHandler) { this.crmfMessageHandler = crmfMessageHandler; }
+    private CertConfirmMessageHandler certConfirmMessageHandler;
+    @Autowired
+    public void setCertConfirmMessageHandler(CertConfirmMessageHandler certConfirmMessageHandler) { this.certConfirmMessageHandler = certConfirmMessageHandler; }
+    private RevocationMessageHandler revocationMessageHandler;
+    @Autowired
+    public void setRevocationMessageHandler(RevocationMessageHandler revocationMessageHandler) { this.revocationMessageHandler = revocationMessageHandler; }
+
+    // -- VALIDATORS
+    @Autowired private HeaderValidator headerValidator;
+    @Autowired private BodyValidator bodyValidator;
+    @Autowired private ProtectionValidator protectionValidator;
+
+    // -- OTHERS
+    private AttributeEngine attributeEngine;
+    @Autowired
+    public void setAttributeEngine(AttributeEngine attributeEngine) {
+        this.attributeEngine = attributeEngine;
+    }
+
+    @Override
+    public ResponseEntity<Object> handlePost(String profileName, byte[] request) throws CmpBaseException {
+        boolean verbose = true;
+
+        init(profileName);
+        validateProfile(profileName);
+
+        final PKIMessage pkiRequest;
+        try { pkiRequest = PKIMessage.getInstance(request); }
+        catch (IllegalArgumentException e) {
+            LOG.error("PN={} | request message cannot be parsed", profileName, e);
+            return buildBadRequest(PkiMessageError.unprotectedMessage(
+                    PKIFailureInfo.badRequest,
+                    ImplFailureInfo.CMPSRV100));
+        }
+        String requestAsString = PkiMessageDumper.dumpPkiMessage(pkiRequest);
+        String logPrefix = PkiMessageDumper.logPrefix(pkiRequest, profileName);
+        LOG.info("{} | request processing: {}",
+                logPrefix,
+                PkiMessageDumper.dumpPkiMessage(verbose, pkiRequest));
+        if(verbose) LOG.info("{} | request as base64: {}",
+                logPrefix,
+                Base64.getEncoder().encodeToString(request));
+
+        // -- (processing) part
+        ConfigurationContext config3gppProfile = new Mobile3gppProfileContext(cmpProfile, pkiRequest,

```

Comment:
What if I do not want to use CMP according to the 3GPP profile?

Reply:
DONE, new cmp_profile column, VARIANT (it is a enum,)